lib/cloud.proxy.js

```diff
@@ -159,125 +209,152 @@ IoTKitCloud.prototype.setDeviceCredentials = function() {
 
 IoTKitCloud.prototype.update = function(callback) {
     var me = this;
-    msg.metadataExtended(me.gatewayId , function (doc) {
-        if(me.deviceName) {
-            doc.name = me.deviceName;
-        }
-        doc.deviceToken = me.secret.deviceToken;
-        doc.deviceId = me.deviceId;
-        me.logger.info("Updating metadata...");
 
-        if(proxyConnector.type === "rest") {
-            //get device to read existing attributes
-            me.proxy.getDevice(doc, function(result) {
+    var handler = function() {
+        msg.metadataExtended(me.gatewayId , function (doc) {
+            if(me.deviceName) {
+                doc.name = me.deviceName;
+            }
+            doc.deviceToken = me.secret.deviceToken;
+            doc.deviceId = me.deviceId;
+            me.logger.info("Updating metadata...");
+
+            if(proxyConnector.type === "rest") {
+                //get device to read existing attributes
+                me.proxy.getDevice(doc, function(result) {
 
-                //append custom attributes to update body
-                for (var attributeName in result.attributes) {
+                    //append custom attributes to update body
+                    for (var attributeName in result.attributes) {
 
-                    if(doc.attributes[attributeName] === undefined) {
-                        doc.attributes[attributeName] = result.attributes[attributeName];
+                        if(doc.attributes[attributeName] === undefined) {
+                            doc.attributes[attributeName] = result.attributes[attributeName];
+                        }
                     }
-                }
-                //update attributes
+                    //update attributes
+                    me.updateAttributes(doc, callback);
+                    //sync componentlist from cloud with local cache
+                    me.updateComponents(result.components)
+                });
+            } else {
                 me.updateAttributes(doc, callback);
-                //sync componentlist from cloud with local cache
-                me.updateComponents(result.components)
-            });
-        } else {
-            me.updateAttributes(doc, callback);
-        }
-        me.logger.info("Metadata updated.");
+            }
+            me.logger.info("Metadata updated.");
 
-    });
+        });
+    };
 
+    me.ensureFreshness(handler);
 };
 
 IoTKitCloud.prototype.updateAttributes = function(doc, callback) {
     var me = this;
-    me.proxy.attributes(doc, function () {
-        me.logger.debug("Attributes returned from " + me.proxy.type);
-        if (callback) {
-            callback();
-        }
-    });
+
+    var handler = function() {
+        me.proxy.attributes(doc, function () {
+            me.logger.debug("Attributes returned from " + me.proxy.type);
+            if (callback) {
+                callback();
+            }
+        });
+    };
+
+    me.ensureFreshness(handler);
 };
```

Comment:
Is it necessary to change this function given it is only ever called from the the update() function and we are checking the device token there first?